<a href="https://colab.research.google.com/github/brianbalb/Consumer_CampaignAnalysis/blob/main/mediaCampaignAnalysis_DataPrep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Import dependences and libraries
import pandas as pd
from google.cloud import bigquery

#Allow connectioin between colab and BigQuery
from google.colab import auth
auth.authenticate_user()

In [ ]:
#Set up BigQuery Cleint in Python
##Set project ID
project_id = "mediacampaignanalysis2025"

#Initialize BigQuery client
client = bigquery.Client(project=project_id)
print("Connected to BQ")

Connected to BQ


In [ ]:
# Ensure correct project ID, dataset, and table
client = bigquery.Client(project="mediacampaignanalysis2025")

datasets = list(client.list_datasets())  # List datasets in the project
print("Datasets:", [dataset.dataset_id for dataset in datasets])

tables = list(client.list_tables("media_campaign_data"))  # List tables in dataset
print("Tables:", [table.table_id for table in tables])


Datasets: ['media_campaign_data']
Tables: ['campaign_performance', 'cleaned_marketing_data']


In [ ]:
#Run Test Query to validate connection and existance of table values
query = """
SELECT COUNT(*) AS total_rows
FROM `mediacampaignanalysis2025.media_campaign_data.campaign_performance`
"""

query_job = client.query(query)
result = query_job.result()

for row in result:
  print(f"Total rows in Table: {row.total_rows}")

Total rows in Table: 2240


In [ ]:
#Load Data from BigQuery into Pandas DataFrame for cleaning and preperation
query = """
SELECT * FROM `mediacampaignanalysis2025.media_campaign_data.campaign_performance`
"""
df = client.query(query).to_dataframe()

# Display first few rows
df.head()


ID  Year_Birth   Education Marital_Status  Income  Kidhome  Teenhome  \
0   8720        1978    2n Cycle       Together     NaN        0         0   
1   6862        1971  Graduation       Divorced  1730.0        0         0   
2   5376        1979  Graduation        Married  2447.0        1         0   
3  11110        1973  Graduation         Single  3502.0        1         0   
4   9931        1963         PhD        Married  4023.0        1         1   

  Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  AcceptedCmp3  \
0  2012-08-12       53        32  ...                  0             0   
1  2014-05-18       65         1  ...                 20             0   
2  2013-01-06       42         1  ...                  1             0   
3  2013-04-13       56         2  ...                 14             0   
4  2014-06-23       29         5  ...                 19             0   

   AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  \
0             1             0             0             0         0   
1             0             0             0             0         0   
2             0             0             0             0         0   
3             0             0             0             0         0   
4             0             0             0             0         0   

   Z_CostContact  Z_Revenue  Response  
0              3         11         0  
1              3         11         0  
2              3         11         0  
3              3         11         0  
4              3         11         0  

[5 rows x 29 columns]

In [ ]:
#Check dataset information and stats
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2240 non-null   Int64  
 1   Year_Birth           2240 non-null   Int64  
 2   Education            2240 non-null   object 
 3   Marital_Status       2240 non-null   object 
 4   Income               2216 non-null   float64
 5   Kidhome              2240 non-null   Int64  
 6   Teenhome             2240 non-null   Int64  
 7   Dt_Customer          2240 non-null   dbdate 
 8   Recency              2240 non-null   Int64  
 9   MntWines             2240 non-null   Int64  
 10  MntFruits            2240 non-null   Int64  
 11  MntMeatProducts      2240 non-null   Int64  
 12  MntFishProducts      2240 non-null   Int64  
 13  MntSweetProducts     2240 non-null   Int64  
 14  MntGoldProds         2240 non-null   Int64  
 15  NumDealsPurchases    2240 non-null   I

ID   Year_Birth         Income   Kidhome  Teenhome    Recency  \
count       2240.0       2240.0    2216.000000    2240.0    2240.0     2240.0   
mean   5592.159821  1968.805804   52247.251354  0.444196   0.50625  49.109375   
std    3246.662198    11.984069   25173.076661  0.538398  0.544538  28.962453   
min            0.0       1893.0    1730.000000       0.0       0.0        0.0   
25%        2828.25       1959.0   35303.000000       0.0       0.0       24.0   
50%         5458.5       1970.0   51381.500000       0.0       0.0       49.0   
75%        8427.75       1977.0   68522.000000       1.0       1.0       74.0   
max        11191.0       1996.0  666666.000000       2.0       2.0       99.0   

         MntWines  MntFruits  MntMeatProducts  MntFishProducts  ...  \
count      2240.0     2240.0           2240.0           2240.0  ...   
mean   303.935714  26.302232           166.95        37.525446  ...   
std    336.597393  39.773434       225.715373        54.628979  ...   
min           0.0        0.0              0.0              0.0  ...   
25%         23.75        1.0             16.0              3.0  ...   
50%         173.5        8.0             67.0             12.0  ...   
75%        504.25       33.0            232.0             50.0  ...   
max        1493.0      199.0           1725.0            259.0  ...   

       NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
count             2240.0        2240.0        2240.0        2240.0   
mean            5.316518      0.072768      0.074554      0.072768   
std             2.426645      0.259813      0.262728      0.259813   
min                  0.0           0.0           0.0           0.0   
25%                  3.0           0.0           0.0           0.0   
50%                  6.0           0.0           0.0           0.0   
75%                  7.0           0.0           0.0           0.0   
max                 20.0           1.0           1.0           1.0   

       AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  \
count        2240.0        2240.0    2240.0         2240.0     2240.0   
mean       0.064286      0.013393  0.009375            3.0       11.0   
std        0.245316      0.114976  0.096391            0.0        0.0   
min             0.0           0.0       0.0            3.0       11.0   
25%             0.0           0.0       0.0            3.0       11.0   
50%             0.0           0.0       0.0            3.0       11.0   
75%             0.0           0.0       0.0            3.0       11.0   
max             1.0           1.0       1.0            3.0       11.0   

       Response  
count    2240.0  
mean   0.149107  
std    0.356274  
min         0.0  
25%         0.0  
50%         0.0  
75%         0.0  
max         1.0  

[8 rows x 26 columns]

In [ ]:
# Check for missing values
df.isnull().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

✅ Observations from Dataset:
Total Rows: 2240
Total Columns: 29
Missing Values: Income column has 24 missing values.
Data Types:
Numerical: (Income, Recency, MntWines, NumWebVisitsMonth, etc.)
Categorical: (Education, Marital_Status)
Date Field: (Dt_Customer needs conversion to datetime format)

Issues to Fix:
The Income column has missing values.
The Year_Birth column might have outliers (ages too old/young).
Some categorical columns need encoding.


📌 Overview: Data Cleaning & Preprocessing
Plan of action details:

✅ Step 1: Assure all columns are numeric for analysis

✅ Step 2: Handle Missing Values (Income)
Since only 24 rows have missing Income, we can:

Fill them with the median income (preferred to avoid skewing data).
Drop not useful.

✅ Step 3: Convert Date Column (Dt_Customer)
convert Dt_Customer to datetime format.
Create Customer_for_Days field.

Why?

Customer_for_Days tells how long the customer has been with the company in days since they joined/purchased.

✅ Step 4: Remove Outliers (Year_Birth)
remove unrealistic ages (e.g., before 1920 or after 2000).

✅ Step 5: Create New Features
1.Total Amount Spent by Customer
2.Total Purchases Across Channels
3.Engagement Rate (Website Visits / Total Purchases)

✅ Step 6: Encode Categorical Variables
Convert Education and Marital_Status into numerical values.

📌Final Step: Save the Cleaned Data
store this cleaned dataset in BigQuery for future use.

In [ ]:
# Convert relevant columns to numeric
columns_to_sum = ["MntWines","MntFruits","MntMeatProducts","MntFishProducts","MntSweetProducts","MntGoldProds","NumDealsPurchases","NumWebPurchases","NumCatalogPurchases","NumStorePurchases","NumWebVisitsMonth"]

for col in columns_to_sum:
    df[col] = pd.to_numeric(df[col], errors="coerce")  # Convert to numeric, set invalid values as NaN

# Now perform the summation correctly
df["Total_Amount_Spent"] = df[columns_to_sum].sum(axis=1)

# verify results
print(df[["Total_Amount_Spent"]].head())  # Check if values are correct


   Total_Amount_Spent
0                1680
1                  43
2                1774
3                  19
4                  43


In [ ]:
print(df.dtypes)  # Check data types of all columns
print(df.head())  # Preview the first few rows


ID                       Int64
Year_Birth               Int64
Education               object
Marital_Status          object
Income                 float64
Kidhome                  Int64
Teenhome                 Int64
Dt_Customer             dbdate
Recency                  Int64
MntWines                 Int64
MntFruits                Int64
MntMeatProducts          Int64
MntFishProducts          Int64
MntSweetProducts         Int64
MntGoldProds             Int64
NumDealsPurchases        Int64
NumWebPurchases          Int64
NumCatalogPurchases      Int64
NumStorePurchases        Int64
NumWebVisitsMonth        Int64
AcceptedCmp3             Int64
AcceptedCmp4             Int64
AcceptedCmp5             Int64
AcceptedCmp1             Int64
AcceptedCmp2             Int64
Complain                 Int64
Z_CostContact            Int64
Z_Revenue                Int64
Response                 Int64
Total_Amount_Spent       Int64
dtype: object
      ID  Year_Birth   Education Marital_Status  Income 

In [ ]:
#Fill missing income values with median values
df.fillna({"Income": df["Income"].median()}, inplace=True)
#Verify
print("missing values handled: ", df.isnull().sum())

missing values handled:  ID                     0
Year_Birth             0
Education              0
Marital_Status         0
Income                 0
Kidhome                0
Teenhome               0
Dt_Customer            0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
AcceptedCmp1           0
AcceptedCmp2           0
Complain               0
Z_CostContact          0
Z_Revenue              0
Response               0
Total_Amount_Spent     0
dtype: int64


In [ ]:
# Convert Dt_Customer to datatime
df["Dt_Customer"] = pd.to_datetime(df["Dt_Customer"])

#Extract useful data-based feature
df["Customer_for_Days"] = (pd.to_datetime("today") - df["Dt_Customer"]).dt.days

#Verify
print("Date column converted")
df[["Dt_Customer", "Customer_for_Days"]].head()

Date column converted


Dt_Customer  Customer_for_Days
0  2012-08-12               4604
1  2014-05-18               3960
2  2013-01-06               4457
3  2013-04-13               4360
4  2014-06-23               3924

In [ ]:
#Define valid age range
df = df[(df["Year_Birth"] >= 1928) & (df["Year_Birth"] <= 2000)]

#verify
print(f"Outliers removed. Reamianing records: {len(df)}")

Outliers removed. Reamianing records: 2237


In [ ]:
#Create new features for future analytical use

##Total amount spent by customer
df["Total_Spending"] = df[
    ["MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds"]
].sum(axis=1)

#verify
print(df[["Total_Spending"]].head())



   Total_Spending
0            1679
1               8
2            1730
3               5
4               9


In [ ]:
##Total Purchases Across Channels
df["Total_Purchases"] = df[
    ["NumDealsPurchases", "NumWebPurchases", "NumCatalogPurchases", "NumStorePurchases"]
].sum(axis=1)

#verify
df["Total_Purchases"].head()

0     1
1    15
2    43
3     0
4    15
Name: Total_Purchases, dtype: Int64

In [ ]:
##Engagement Rate(Website visits/Total Purchases)
df["Engagement_Rate"] = df["NumWebVisitsMonth"] / (df["Total_Purchases"] + 1)

#verify
df["Engagement_Rate"].head()

0         0.0
1        1.25
2    0.022727
3        14.0
4      1.1875
Name: Engagement_Rate, dtype: Float64

In [ ]:
#Encode Categorical Variables Education and Marital_Status into numerial values for ML analysis
df = pd.get_dummies(df, columns=['Education_Basic','Education_Graduation', 'Education_Master', 'Education_PhD','Marital_Status_Alone', 'Marital_Status_Divorced','Marital_Status_Married', 'Marital_Status_Single','Marital_Status_Together', 'Marital_Status_Widow','Marital_Status_YOLO'], drop_first=True)


KeyError: "None of [Index(['Education_Basic', 'Education_Graduation', 'Education_Master',\n       'Education_PhD', 'Marital_Status_Alone', 'Marital_Status_Divorced',\n       'Marital_Status_Married', 'Marital_Status_Single',\n       'Marital_Status_Together', 'Marital_Status_Widow',\n       'Marital_Status_YOLO'],\n      dtype='object')] are in the [columns]"

In [ ]:
# Rename columns to remove "_True"
df.columns = df.columns.str.replace("_True", "")

# Convert boolean values to integers (0/1)
df = df.astype(int)

#verify
print("✅ Column names cleaned and encoded variables converted to 0/1!")
df.head()

ValueError: invalid literal for int() with base 10: '2n Cycle'

FINAL STEP: Save the cleaned Data

In [ ]:
table_id = "mediacampaignanalysis2025.media_campaign_data.cleaned_marketing_data"

#Upload cleaned Dataframe to BigQuery
job = client.load_table_from_dataframe(df, table_id)
job.result()

#verify
print("Cleaned data uploaded to BigQuery")

ArrowInvalid: Float value 51381.5 was truncated converting to int64

In [ ]:
# Load cleaned dataset from BigQuery (or use ABM simulation results)
query = "SELECT * FROM `mediacampaignanalysis2025.media_campaign_data.cleaned_marketing_data`"
df = client.query(query).to_dataframe()
drive.mount('/content/drive')

# Save dataset to Excel
file_path = "/content/drive"
df.to_excel(file_path, index=False)

print("✅ Excel file saved successfully!")


OSError: Cannot save file into a non-existent directory: '/drive/my-drive'